cloud-storage/cloud_storage.c

```diff
@@ -0,0 +1,947 @@
+      {
+         scond_wait(_operation_condition, _operation_queue_mutex);
+      }
+
+      if (_shutdown)
+      {
+         slock_unlock(_operation_queue_mutex);
+         continue;
+      }
+
+      operation = (struct _operation_t *)generic_queue_pop(_operation_queue);
+      slock_unlock(_operation_queue_mutex);
+
+      if (operation)
+      {
+         switch (operation->operation_type)
+         {
+            case SYNC_FILES:
+               _sync_files_execute(operation->parameters.sync.folder_type);
+               break;
+            case UPLOAD_FILE:
+               _upload_file_execute(operation->parameters.upload.folder_type, operation->parameters.upload.local_file);
+               break;
+            case DOWNLOAD_FILE:
+               break;
+         }
+
+         _free_operation(operation);
+      }
+   }
+
+   /* Broke out of the operation loop, will shutdown the operation thread */
+   if (_operation_queue)
+   {
+      generic_queue_free(_operation_queue, &_free_operation_fn);
+   }
+}
+
+void cloud_storage_init(void)
+{
+   struct _operation_queue_item *sync_item;
+
+   if (!_shutdown)
+   {
+      return;
+   }
+
+   _provider = cloud_storage_local_folder_create();
+
+   _operation_queue = generic_queue_new();
+   _operation_queue_mutex = slock_new();
+   _operation_condition = scond_new();
+
+   _shutdown = false;
+   _operation_thread = sthread_create(_operation_thread_loop, NULL);
+
+   if (_provider->ready_for_request())
+   {
+      cloud_storage_sync_files();
+   }
+}
+
+void cloud_storage_shutdown(void)
+{
+   if (_shutdown)
+   {
+      return;
+   }
+
+   /* Signal the operation thread to shutdown in a critical section */
+   slock_lock(_operation_queue_mutex);
+   _shutdown = true;
+   scond_signal(_operation_condition);
+   slock_unlock(_operation_queue_mutex);
+
+   sthread_join(_operation_thread);
+}
+
+char *cloud_storage_get_md5_hash(char *absolute_filename)
+{
+   RFILE *file;
+   uint8_t buffer[BUFFER_SIZE];
+   MD5_CTX md5_ctx;
+   int64_t bytes_read;
+   unsigned char *checksum;
+   char *checksum_str;
+
+   file = filestream_open(absolute_filename, RETRO_VFS_FILE_ACCESS_READ, RETRO_VFS_FILE_ACCESS_HINT_NONE);
+   if (!file)
+   {
+      return NULL;
+   }
+
+   MD5_Init(&md5_ctx);
+
+   /* Pass the contents of the local file through the MD5 hash function */
+   bytes_read = filestream_read(file, buffer, BUFFER_SIZE);
+   while (bytes_read > 0)
+   {
+      MD5_Update(&md5_ctx, buffer, bytes_read);
+      bytes_read = filestream_read(file, buffer, BUFFER_SIZE);
+   }
+
+   filestream_close(file);
+   checksum = (unsigned char *)calloc(16, sizeof(char));
+   MD5_Final(checksum, &md5_ctx);
+
+   /* Convert the MD5 checksum to a hex string */
+   checksum_str = bytes_to_hex_str(checksum, 16);
+   free(checksum);
+
+   return checksum_str;
+}
+
+/* Converts a value (game save/state directory setting) to an absolute directory path */
+static char *_config_value_to_directory(const char *value)
```

Comment:
Could it be that the "fill_pathname_parent_dir_name" function does the same as this function?
IIUC you are getting the basedir of a file/dir without the trailing slash right?

Reply:
The function "fill_pathname_parent_dir_name" gets the name of the immediate parent directory. I want the absolute path minus the last path component.
This has been changed to use "fill_pathname_basedir".